File: quick_sort_pre/quick_sort.rb
def quick_sort(arr)
  return arr if arr.length <= 1

  pivot, left, right = arr[0], [], []

  arr.each_with_index do |el, idx|
    next if idx == 0
    pivot > el ? left.push(el) : right.push(el)
  end

  quick_sort(left) + [pivot] + quick_sort(right)
end


# test
arr = [1,2,3,4,5]

p quick_sort(arr) == arr
p quick_sort(arr.shuffle) == arr
p quick_sort(arr.reverse) == arr


# [3,7,2,5,1]

# [1,2,3,5,7]

# pivot = 3
# idx = 0

# pointer = 1
# barrier = 1

# if pivot > curr_el
#   swap curr_el with the first el after barrier
# end

File: quick_sort_pre/quicksort.js
function quicksort(arr) {
  if (arr.length === 0) {
    return []
  } else if (arr.length === 1) {
    return arr
  }

  const pivot = arr[0]

  const leftArr = []
  const rightArr = []

  arr.splice(1).forEach(item => {
    if (item < pivot) {
      leftArr.push(item)
    } else {
      rightArr.push(item)
    }
  })

  const result = [...quicksort(leftArr), pivot, ...quicksort(rightArr)]

  return result
}

const arr = [3, 7, 5, 2, 1]
console.log(quicksort(arr));